### PR TITLE
Ingester: test hitting per-instance limits with circuit breaker

### PR DESCRIPTION
**Pull request by duricanikolic**
#### What this PR does
This PR introduces a test that shows that, when a circuit breaker is setup for ingesters, and when an ingester hits a per-instance limit, the circuit breaker is triggered.

Circuit breakers are introduced in this [PR](https://github.com/grafana/mimir/pull/5650).

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by pstibrany** [comment]
It seems to me that `circuitbreaker_test.go` already tests circuit breaker behaviour in the ingester client. This test is extended version of that, also testing for instance limits.

I'm not sure if this test adds much value in this already gigantic test file, but it runs quickly, so I don't mind having it. Let's wait for @56quarters opinion.

**Review by 56quarters — COMMENTED** [review]
This is starting an entire ingester server and making calls to it over gRPC which seems like it's really straining the definition of a "unit" test. Like Peter mentioned, the behavior of the circuit breaker is already tested elsewhere.

I'd rather see these things tested at a smaller scope:
* Test that the errors returned by an ingester are correctly built (wrapping what we expect and implementing the various interfaces we expect)
* Test that the ingester client uses the circuit breaking interceptor correctly

... But I don't feel particularly strongly about it. If you thing there's value in testing this here, that's fine with me.

File: pkg/ingester/ingester_test.go [hunk 1]
```diff
@@ -5889,6 +5890,207 @@ func TestIngester_PushInstanceLimits(t *testing.T) {
...
+						// The first request causes an instance limit error,
+						// and which opens the circuit breaker, but no
+						// circuitbreaker.ErrCircuitBreakerOpen is returned.
+						_, err := client.Push(ctx, req)
+						require.Error(t, err)
+						require.NotErrorIs(t, err, circuitbreaker.ErrCircuitBreakerOpen)
+						s, ok := status.FromError(err)
+						require.True(t, ok, "expected to be able to convert to gRPC status")
+						assert.Equal(t, codes.Unavailable, s.Code())
+						require.ErrorContains(t, s.Err(), testData.expectedErr.Error())
```
[the remark this answers is not in the document]
Comment (duricanikolic) [reply pkg/ingester/ingester_test.go:6062]: @56quarters `test.expectedErr` is the error that the server side is supposed to return (`IngesterServer`), while the error obtained here is the one obtained from the client. They are compatible in the sense that they do have the same gRPC code and the same error message, but they are not the same object.

**Comment by colega** [comment]
I think testing the integration between the client and the actual ingester here is the best approach, as that's what important here. We already know that circuitbreaker works, we need to check that it works in all our error mapping over grpc mess.

The issue was that @duricanikolic called it a _unit_ test, which is not.

**Comment by duricanikolic** [comment]
> This is starting an entire ingester server and making calls to it over gRPC which seems like it's really straining the definition of a "unit" test. Like Peter mentioned, the behavior of the circuit breaker is already tested elsewhere.
> 
> I'd rather see these things tested at a smaller scope:
> 
> * Test that the errors returned by an ingester are correctly built (wrapping what we expect and implementing the various interfaces we expect)
> * Test that the ingester client uses the circuit breaking interceptor correctly
> 
> ... But I don't feel particularly strongly about it. If you thing there's value in testing this here, that's fine with me.

Hi @56quarters, 
I agree that this test doesn't respect the definition of a unit test, but there are different other tests in the same source doing something similar. 

> * Test that the errors returned by an ingester are correctly built (wrapping what we expect and implementing the various interfaces we expect)

This is already tested in `TestIngester_PushInstanceLimits`. The goal of the new test is:
- to show that the first time an instance limit is hit, the correlated error is returned, and the circuit breaker is opened
- the second time the same limit is hit, the correlated error is not returned, but `circuitbreaker.ErrCircuitBreakerOpen` is
- that the `cortex_ingester_client_circuit_breaker_results_total` counter gets correctly updated during the scenarios mentioned above

**Comment by duricanikolic** [comment]
> The issue was that @duricanikolic called it a _unit_ test, which is not.
I agree it is not a unit test. It is more a functional test.

**Pull request merged**